In [3]:
# (1) 필요한 모듈 import
from sklearn.datasets import load_iris #싸이킷런 데이터셋에 있는 iris를 불러오는 코드
from sklearn.model_selection import train_test_split #싸이킷런에 있는 데이터를 train과 test dataset으로 나누어주는 함수 import
from sklearn.tree import DecisionTreeClassifier # 싸이킷런에 있는 결정트리분류기를 사용하기 위해 불러오는 코드
from sklearn.metrics import classification_report #싸이킷런에 있는 분류 결과에 대한 시각화를 위해 쓰는 코드

# (2) 데이터 준비
iris = load_iris() #iris 데이터 전체를 불러온다.
iris_data = iris.data #iris데이터의 data컬럼을 분류해 iris_data 변수에 담는다.
iris_label = iris.target #iris데이터의 target컬럼을 분류해 iris_label 변수에 담는다.

# (3) train, test 데이터 분리
#train_test_split()를 사용하여 X값, y값을 각각 train data와 test data로 나눈다. 함수에 들어 갈 파라미터로는 x,y가 들어가고
#test_size는 몇대몇으로 나눌지 정하는 옵션, random_state는 랜덤 패턴의 값을 지정한다. (어떤 값을 넣어도 무방하다.)
X_train, X_test, y_train, y_test = train_test_split(iris_data, 
                                                    iris_label, 
                                                    test_size=0.2, 
                                                    random_state=7) 


# (4) 모델 학습 및 예측
decision_tree = DecisionTreeClassifier(random_state=32) #결정트리분류기의 객체를 만든다.
decision_tree.fit(X_train, y_train) # 분류기에 x와 y의 훈련 데이터를 넣어 훈련 시킨다.
y_pred = decision_tree.predict(X_test) # 훈련된 분류기에 X_test라는 테스트 데이터셋을 넣어 얼마나 예측했는지 확인한다.

print(classification_report(y_test, y_pred)) # 결과를 지표로 확인하기 위해 classification_report를 활용해 y_test, y_pred 값을 넣어 확인한다.


              precision    recall  f1-score   support

           0       1.00      1.00      1.00         7
           1       0.91      0.83      0.87        12
           2       0.83      0.91      0.87        11

    accuracy                           0.90        30
   macro avg       0.91      0.91      0.91        30
weighted avg       0.90      0.90      0.90        30



In [2]:
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from sklearn.tree import DecisionTreeClassifier
import pandas as pd
from scipy import stats

digits = load_digits()
digits_data = digits.data
digits_label = digits.target
print(digits.target)
print(digits.DESCR)

X_train, X_test, y_train, y_test = train_test_split(digits_data, digits_label, 
                                                    test_size=0.2, random_state=1)

classifier = DecisionTreeClassifier(random_state=32)
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)

print(classification_report(y_test, y_pred))

[0 1 2 ... 8 9 8]
.. _digits_dataset:

Optical recognition of handwritten digits dataset
--------------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 1797
    :Number of Attributes: 64
    :Attribute Information: 8x8 image of integer pixels in the range 0..16.
    :Missing Attribute Values: None
    :Creator: E. Alpaydin (alpaydin '@' boun.edu.tr)
    :Date: July; 1998

This is a copy of the test set of the UCI ML hand-written digits datasets
https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits

The data set contains images of hand-written digits: 10 classes where
each class refers to a digit.

Preprocessing programs made available by NIST were used to extract
normalized bitmaps of handwritten digits from a preprinted form. From a
total of 43 people, 30 contributed to the training set and different 13
to the test set. 32x32 bitmaps are divided into nonoverlapping blocks of
4x4 and the number of on pixels are c

In [5]:
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import SGDClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier

classifier = DecisionTreeClassifier()
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)

print(classification_report(y_test, y_pred), '<Decision Tree>')


classifier = SGDClassifier()
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)

print(classification_report(y_test, y_pred), '<SGD Classifier>')


classifier = SVC()
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)

print(classification_report(y_test, y_pred), '<SVM>')

classifier = LogisticRegression()
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)

print(classification_report(y_test, y_pred), '<Logistic Regression>')


classifier = RandomForestClassifier()
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)

print(classification_report(y_test, y_pred), '<Random Forest Classifier>')


              precision    recall  f1-score   support

           0       1.00      0.93      0.96        43
           1       0.93      0.80      0.86        35
           2       0.91      0.86      0.89        36
           3       0.81      0.73      0.77        41
           4       0.84      0.84      0.84        38
           5       0.76      0.97      0.85        30
           6       0.92      0.97      0.95        37
           7       0.86      0.86      0.86        37
           8       0.81      0.86      0.83        29
           9       0.72      0.76      0.74        34

    accuracy                           0.86       360
   macro avg       0.86      0.86      0.86       360
weighted avg       0.86      0.86      0.86       360
 <Decision Tree>
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.97      0.94      0.96        35
           2       1.00      1.00      1.00        36
        

C:\Users\JD-Laptop\AppData\Roaming\Python\Python39\site-packages\sklearn\linear_model\_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       1.00      1.00      1.00        35
           2       1.00      1.00      1.00        36
           3       1.00      0.93      0.96        41
           4       0.97      1.00      0.99        38
           5       0.97      1.00      0.98        30
           6       1.00      1.00      1.00        37
           7       1.00      0.97      0.99        37
           8       0.94      1.00      0.97        29
           9       0.94      0.97      0.96        34

    accuracy                           0.98       360
   macro avg       0.98      0.98      0.98       360
weighted avg       0.98      0.98      0.98       360
 <Random Forest Classifier>


In [8]:
!pip install lazypredict
from lazypredict.Supervised import LazyClassifier
clf = LazyClassifier(verbose=0,ignore_warnings=True, custom_metric=None)
models,predictions = clf.fit(X_train, X_test, y_train, y_test)
models

100%|███████████████████████████████████████████████████████| 29/29 [00:07<00:00,  4.12it/s]


Accuracy  Balanced Accuracy ROC AUC  F1 Score  \
Model                                                                          
ExtraTreesClassifier               0.99               0.99    None      0.99   
RandomForestClassifier             0.98               0.98    None      0.98   
SVC                                0.98               0.98    None      0.98   
LogisticRegression                 0.98               0.98    None      0.98   
LGBMClassifier                     0.98               0.98    None      0.98   
KNeighborsClassifier               0.98               0.98    None      0.98   
XGBClassifier                      0.97               0.97    None      0.97   
CalibratedClassifierCV             0.97               0.97    None      0.97   
LinearSVC                          0.97               0.97    None      0.97   
PassiveAggressiveClassifier        0.96               0.96    None      0.96   
LinearDiscriminantAnalysis         0.96               0.96    None      0.96   
LabelSpreading                     0.96               0.96    None      0.96   
LabelPropagation                   0.96               0.96    None      0.96   
BaggingClassifier                  0.96               0.96    None      0.96   
NuSVC                              0.96               0.95    None      0.96   
RidgeClassifierCV                  0.95               0.95    None      0.95   
RidgeClassifier                    0.95               0.95    None      0.95   
Perceptron                         0.94               0.94    None      0.94   
SGDClassifier                      0.94               0.94    None      0.94   
NearestCentroid                    0.88               0.88    None      0.88   
BernoulliNB                        0.87               0.86    None      0.87   
DecisionTreeClassifier             0.85               0.85    None      0.85   
QuadraticDiscriminantAnalysis      0.84               0.84    None      0.84   
ExtraTreeClassifier                0.81               0.81    None      0.81   
GaussianNB                         0.79               0.79    None      0.79   
AdaBoostClassifier                 0.26               0.24    None      0.19   
DummyClassifier                    0.08               0.10    None      0.01   

                               Time Taken  
Model                                      
ExtraTreesClassifier                 0.43  
RandomForestClassifier               0.47  
SVC                                  0.19  
LogisticRegression                   0.11  
LGBMClassifier                       1.30  
KNeighborsClassifier                 0.15  
XGBClassifier                        0.81  
CalibratedClassifierCV               1.36  
LinearSVC                            0.28  
PassiveAggressiveClassifier          0.08  
LinearDiscriminantAnalysis           0.06  
LabelSpreading                       0.26  
LabelPropagation                     0.22  
BaggingClassifier                    0.19  
NuSVC                                0.34  
RidgeClassifierCV                    0.03  
RidgeClassifier                      0.03  
Perceptron                           0.06  
SGDClassifier                        0.10  
NearestCentroid                      0.02  
BernoulliNB                          0.03  
DecisionTreeClassifier               0.05  
QuadraticDiscriminantAnalysis        0.04  
ExtraTreeClassifier                  0.03  
GaussianNB                           0.02  
AdaBoostClassifier                   0.32  
DummyClassifier                      0.02

In [16]:
from sklearn.ensemble import ExtraTreesClassifier
classifier = ExtraTreesClassifier(n_estimators = 100)
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       1.00      1.00      1.00        35
           2       1.00      1.00      1.00        36
           3       1.00      1.00      1.00        41
           4       0.97      1.00      0.99        38
           5       0.97      1.00      0.98        30
           6       1.00      1.00      1.00        37
           7       1.00      0.97      0.99        37
           8       1.00      1.00      1.00        29
           9       0.97      0.97      0.97        34

    accuracy                           0.99       360
   macro avg       0.99      0.99      0.99       360
weighted avg       0.99      0.99      0.99       360



In [ ]:
#--------------------------------------------------------------------------

In [3]:
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

wine = load_wine()
wine_data = wine.data
wine_label = wine.target
print(wine.target)
print(wine.DESCR)

X_train, X_test, y_train, y_test = train_test_split(wine_data, wine_label, 
                                                    test_size=0.2, random_state=1)

classifier = DecisionTreeClassifier(random_state=32)
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)

print(classification_report(y_test, y_pred))

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2]
.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= ====

In [39]:
classifier = DecisionTreeClassifier()
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)

print(classification_report(y_test, y_pred), '<Decision Tree>')


classifier = SGDClassifier()
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)

print(classification_report(y_test, y_pred), '<SGD Classifier>')


classifier = SVC()
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)

print(classification_report(y_test, y_pred), '<SVM>')

classifier = LogisticRegression()
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)

print(classification_report(y_test, y_pred), '<Logistic Regression>')


classifier = RandomForestClassifier()
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)

print(classification_report(y_test, y_pred), '<Random Forest Classifier>')

              precision    recall  f1-score   support

           0       0.93      0.93      0.93        14
           1       0.75      0.92      0.83        13
           2       1.00      0.67      0.80         9

    accuracy                           0.86        36
   macro avg       0.89      0.84      0.85        36
weighted avg       0.88      0.86      0.86        36
 <Decision Tree>
              precision    recall  f1-score   support

           0       1.00      0.86      0.92        14
           1       0.54      1.00      0.70        13
           2       0.00      0.00      0.00         9

    accuracy                           0.69        36
   macro avg       0.51      0.62      0.54        36
weighted avg       0.58      0.69      0.61        36
 <SGD Classifier>
              precision    recall  f1-score   support

           0       0.92      0.79      0.85        14
           1       0.58      0.85      0.69        13
           2       0.20      0.11      0.1

In [40]:
clf = LazyClassifier(verbose=0,ignore_warnings=True, custom_metric=None)
models,predictions = clf.fit(X_train, X_test, y_train, y_test)
models

100%|███████████████████████████████████████████████████████| 29/29 [00:00<00:00, 30.36it/s]


Accuracy  Balanced Accuracy ROC AUC  F1 Score  \
Model                                                                          
LinearSVC                          1.00               1.00    None      1.00   
LinearDiscriminantAnalysis         1.00               1.00    None      1.00   
RidgeClassifierCV                  1.00               1.00    None      1.00   
RidgeClassifier                    1.00               1.00    None      1.00   
QuadraticDiscriminantAnalysis      1.00               1.00    None      1.00   
PassiveAggressiveClassifier        1.00               1.00    None      1.00   
NuSVC                              1.00               1.00    None      1.00   
LogisticRegression                 1.00               1.00    None      1.00   
LGBMClassifier                     1.00               1.00    None      1.00   
GaussianNB                         1.00               1.00    None      1.00   
KNeighborsClassifier               0.97               0.97    None      0.97   
ExtraTreeClassifier                0.97               0.97    None      0.97   
XGBClassifier                      0.97               0.97    None      0.97   
SGDClassifier                      0.97               0.97    None      0.97   
RandomForestClassifier             0.97               0.97    None      0.97   
LabelPropagation                   0.97               0.97    None      0.97   
Perceptron                         0.97               0.97    None      0.97   
NearestCentroid                    0.97               0.97    None      0.97   
ExtraTreesClassifier               0.97               0.97    None      0.97   
LabelSpreading                     0.97               0.97    None      0.97   
CalibratedClassifierCV             0.97               0.96    None      0.97   
SVC                                0.97               0.96    None      0.97   
BaggingClassifier                  0.94               0.95    None      0.94   
BernoulliNB                        0.92               0.91    None      0.92   
DecisionTreeClassifier             0.86               0.84    None      0.86   
AdaBoostClassifier                 0.83               0.80    None      0.83   
DummyClassifier                    0.36               0.33    None      0.19   

                               Time Taken  
Model                                      
LinearSVC                            0.01  
LinearDiscriminantAnalysis           0.01  
RidgeClassifierCV                    0.02  
RidgeClassifier                      0.01  
QuadraticDiscriminantAnalysis        0.01  
PassiveAggressiveClassifier          0.02  
NuSVC                                0.01  
LogisticRegression                   0.02  
LGBMClassifier                       0.11  
GaussianNB                           0.01  
KNeighborsClassifier                 0.01  
ExtraTreeClassifier                  0.01  
XGBClassifier                        0.08  
SGDClassifier                        0.01  
RandomForestClassifier               0.18  
LabelPropagation                     0.01  
Perceptron                           0.02  
NearestCentroid                      0.01  
ExtraTreesClassifier                 0.14  
LabelSpreading                       0.02  
CalibratedClassifierCV               0.05  
SVC                                  0.01  
BaggingClassifier                    0.03  
BernoulliNB                          0.01  
DecisionTreeClassifier               0.01  
AdaBoostClassifier                   0.10  
DummyClassifier                      0.01

In [44]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
classifier = LinearDiscriminantAnalysis()
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        14
           1       1.00      1.00      1.00        13
           2       1.00      1.00      1.00         9

    accuracy                           1.00        36
   macro avg       1.00      1.00      1.00        36
weighted avg       1.00      1.00      1.00        36



In [ ]:
#---------------------------------------------------------------

In [4]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

breast_cancer = load_breast_cancer()
breast_cancer_data = breast_cancer.data
breast_cancer_label = breast_cancer.target
print(breast_cancer.target)
print(breast_cancer.DESCR)

X_train, X_test, y_train, y_test = train_test_split(breast_cancer_data, breast_cancer_label, 
                                                    test_size=0.2, random_state=1)

classifier = DecisionTreeClassifier(random_state=32)
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)

print(classification_report(y_test, y_pred))

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 1 0 0 0 0 0 0 0 0 1 0 1 1 1 1 1 0 0 1 0 0 1 1 1 1 0 1 0 0 1 1 1 1 0 1 0 0
 1 0 1 0 0 1 1 1 0 0 1 0 0 0 1 1 1 0 1 1 0 0 1 1 1 0 0 1 1 1 1 0 1 1 0 1 1
 1 1 1 1 1 1 0 0 0 1 0 0 1 1 1 0 0 1 0 1 0 0 1 0 0 1 1 0 1 1 0 1 1 1 1 0 1
 1 1 1 1 1 1 1 1 0 1 1 1 1 0 0 1 0 1 1 0 0 1 1 0 0 1 1 1 1 0 1 1 0 0 0 1 0
 1 0 1 1 1 0 1 1 0 0 1 0 0 0 0 1 0 0 0 1 0 1 0 1 1 0 1 0 0 0 0 1 1 0 0 1 1
 1 0 1 1 1 1 1 0 0 1 1 0 1 1 0 0 1 0 1 1 1 1 0 1 1 1 1 1 0 1 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 1 1 1 1 1 1 0 1 0 1 1 0 1 1 0 1 0 0 1 1 1 1 1 1 1 1 1 1 1 1
 1 0 1 1 0 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 0 1 0 1 1 1 1 0 0 0 1 1
 1 1 0 1 0 1 0 1 1 1 0 1 1 1 1 1 1 1 0 0 0 1 1 1 1 1 1 1 1 1 1 1 0 0 1 0 0
 0 1 0 0 1 1 1 1 1 0 1 1 1 1 1 0 1 1 1 0 1 1 0 0 1 1 1 1 1 1 0 1 1 1 1 1 1
 1 0 1 1 1 1 1 0 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 0 1 0 0 1 0 1 1 1 1 1 0 1 1
 0 1 0 1 1 0 1 0 1 1 1 1 1 1 1 1 0 0 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 0 1
 1 1 1 1 1 1 0 1 0 1 1 0 

In [6]:
classifier = DecisionTreeClassifier()
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)

print(classification_report(y_test, y_pred), '<Decision Tree>')


classifier = SGDClassifier()
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)

print(classification_report(y_test, y_pred), '<SGD Classifier>')


classifier = SVC()
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)

print(classification_report(y_test, y_pred), '<SVM>')

classifier = LogisticRegression()
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)

print(classification_report(y_test, y_pred), '<Logistic Regression>')


classifier = RandomForestClassifier()
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)

print(classification_report(y_test, y_pred), '<Random Forest Classifier>')

NameError: name 'X_train' is not defined

In [49]:
clf = LazyClassifier(verbose=0,ignore_warnings=True, custom_metric=None)
models,predictions = clf.fit(X_train, X_test, y_train, y_test)
models

100%|███████████████████████████████████████████████████████| 29/29 [00:01<00:00, 17.81it/s]


Accuracy  Balanced Accuracy  ROC AUC  F1 Score  \
Model                                                                           
LinearSVC                          0.97               0.97     0.97      0.97   
SVC                                0.97               0.97     0.97      0.97   
PassiveAggressiveClassifier        0.97               0.97     0.97      0.97   
LogisticRegression                 0.97               0.97     0.97      0.97   
Perceptron                         0.97               0.96     0.96      0.97   
CalibratedClassifierCV             0.96               0.96     0.96      0.96   
SGDClassifier                      0.96               0.96     0.96      0.96   
BaggingClassifier                  0.96               0.95     0.95      0.96   
QuadraticDiscriminantAnalysis      0.96               0.95     0.95      0.96   
LGBMClassifier                     0.96               0.95     0.95      0.96   
DecisionTreeClassifier             0.96               0.95     0.95      0.96   
KNeighborsClassifier               0.96               0.94     0.94      0.96   
XGBClassifier                      0.96               0.94     0.94      0.96   
ExtraTreesClassifier               0.96               0.94     0.94      0.96   
RandomForestClassifier             0.96               0.94     0.94      0.96   
RidgeClassifier                    0.96               0.94     0.94      0.96   
GaussianNB                         0.95               0.94     0.94      0.95   
NuSVC                              0.95               0.93     0.93      0.95   
LinearDiscriminantAnalysis         0.95               0.93     0.93      0.95   
RidgeClassifierCV                  0.95               0.93     0.93      0.95   
BernoulliNB                        0.94               0.92     0.92      0.94   
LabelPropagation                   0.93               0.92     0.92      0.93   
LabelSpreading                     0.93               0.92     0.92      0.93   
AdaBoostClassifier                 0.93               0.91     0.91      0.93   
ExtraTreeClassifier                0.93               0.90     0.90      0.93   
NearestCentroid                    0.92               0.90     0.90      0.92   
DummyClassifier                    0.63               0.50     0.50      0.49   

                               Time Taken  
Model                                      
LinearSVC                            0.02  
SVC                                  0.03  
PassiveAggressiveClassifier          0.02  
LogisticRegression                   0.02  
Perceptron                           0.02  
CalibratedClassifierCV               0.05  
SGDClassifier                        0.02  
BaggingClassifier                    0.07  
QuadraticDiscriminantAnalysis        0.02  
LGBMClassifier                       0.19  
DecisionTreeClassifier               0.02  
KNeighborsClassifier                 0.10  
XGBClassifier                        0.10  
ExtraTreesClassifier                 0.16  
RandomForestClassifier               0.35  
RidgeClassifier                      0.02  
GaussianNB                           0.01  
NuSVC                                0.04  
LinearDiscriminantAnalysis           0.02  
RidgeClassifierCV                    0.02  
BernoulliNB                          0.01  
LabelPropagation                     0.04  
LabelSpreading                       0.03  
AdaBoostClassifier                   0.20  
ExtraTreeClassifier                  0.01  
NearestCentroid                      0.02  
DummyClassifier                      0.01

In [55]:
classifier = RandomForestClassifier()
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.88      0.94        42
           1       0.94      1.00      0.97        72

    accuracy                           0.96       114
   macro avg       0.97      0.94      0.95       114
weighted avg       0.96      0.96      0.96       114



In [ ]:
# 예상치않게 종료됨.--------------------------------------------------------
# from sklearn.model_selection import GridSearchCV
# import numpy as np

# # Number of trees in random forest
# n_estimators = [int(x) for x in np.linspace(start = 200, stop = 2000, num = 10)]
# # Number of features to consider at every split
# max_features = ['auto', 'sqrt']
# # Maximum number of levels in tree
# max_depth = [int(x) for x in np.linspace(10, 110, num = 11)]
# max_depth.append(None)
# # Minimum number of samples required to split a node
# min_samples_split = [2, 5, 10]
# # Minimum number of samples required at each leaf node
# min_samples_leaf = [1, 2, 4]
# # Method of selecting samples for training each tree
# bootstrap = [True, False]
# # Create the random grid

# random_grid = {'n_estimators': n_estimators,
#                'max_features': max_features,
#                'max_depth': max_depth,
#                'min_samples_split': min_samples_split,
#                'min_samples_leaf': min_samples_leaf,
#                'bootstrap': bootstrap}
# print(random_grid)
# rf = RandomForestClassifier()
# grid_search = GridSearchCV(estimator = rf, param_grid = random_grid, cv = 2, n_jobs = -1, verbose = 2)

# grid_search.fit(X_train, y_train)

{'n_estimators': [200, 400, 600, 800, 1000, 1200, 1400, 1600, 1800, 2000], 'max_features': ['auto', 'sqrt'], 'max_depth': [10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, None], 'min_samples_split': [2, 5, 10], 'min_samples_leaf': [1, 2, 4], 'bootstrap': [True, False]}
Fitting 2 folds for each of 4320 candidates, totalling 8640 fits


In [13]:
n_estimator_lists = [i for i in range(100,2000,100)]
min_samples_leafs_lists = [1, 2, 3, 4, 5, 6, 7]
for n_estimator_list in n_estimator_lists:
    for min_samples_leaf_list in min_samples_leafs_lists:
        classifier = RandomForestClassifier(
            n_estimators=n_estimator_list,
            max_features='auto',
            min_samples_leaf=min_samples_leaf_list)
        print(classification_report(y_test, y_pred))
        print('n_estimator =', n_estimator_list)
        print('min_samples_leaf =', min_samples_leaf_list)

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       1.00      1.00      1.00        35
           2       1.00      1.00      1.00        36
           3       1.00      0.93      0.96        41
           4       0.97      1.00      0.99        38
           5       0.97      1.00      0.98        30
           6       1.00      1.00      1.00        37
           7       1.00      0.97      0.99        37
           8       0.94      1.00      0.97        29
           9       0.94      0.97      0.96        34

    accuracy                           0.98       360
   macro avg       0.98      0.98      0.98       360
weighted avg       0.98      0.98      0.98       360

n_estimator = 100
min_samples_leaf = 1
              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       1.00      1.00      1.00        35
           2       1.00      1.00      

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       1.00      1.00      1.00        35
           2       1.00      1.00      1.00        36
           3       1.00      0.93      0.96        41
           4       0.97      1.00      0.99        38
           5       0.97      1.00      0.98        30
           6       1.00      1.00      1.00        37
           7       1.00      0.97      0.99        37
           8       0.94      1.00      0.97        29
           9       0.94      0.97      0.96        34

    accuracy                           0.98       360
   macro avg       0.98      0.98      0.98       360
weighted avg       0.98      0.98      0.98       360

n_estimator = 500
min_samples_leaf = 6
              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       1.00      1.00      1.00        35
           2       1.00      1.00      

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       1.00      1.00      1.00        35
           2       1.00      1.00      1.00        36
           3       1.00      0.93      0.96        41
           4       0.97      1.00      0.99        38
           5       0.97      1.00      0.98        30
           6       1.00      1.00      1.00        37
           7       1.00      0.97      0.99        37
           8       0.94      1.00      0.97        29
           9       0.94      0.97      0.96        34

    accuracy                           0.98       360
   macro avg       0.98      0.98      0.98       360
weighted avg       0.98      0.98      0.98       360

n_estimator = 1000
min_samples_leaf = 5
              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       1.00      1.00      1.00        35
           2       1.00      1.00     

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       1.00      1.00      1.00        35
           2       1.00      1.00      1.00        36
           3       1.00      0.93      0.96        41
           4       0.97      1.00      0.99        38
           5       0.97      1.00      0.98        30
           6       1.00      1.00      1.00        37
           7       1.00      0.97      0.99        37
           8       0.94      1.00      0.97        29
           9       0.94      0.97      0.96        34

    accuracy                           0.98       360
   macro avg       0.98      0.98      0.98       360
weighted avg       0.98      0.98      0.98       360

n_estimator = 1600
min_samples_leaf = 2
              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       1.00      1.00      1.00        35
           2       1.00      1.00     

In [14]:
classifier = RandomForestClassifier(n_estimators=100)
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.95      0.98        43
           1       1.00      1.00      1.00        35
           2       1.00      0.97      0.99        36
           3       1.00      0.98      0.99        41
           4       0.95      1.00      0.97        38
           5       0.97      0.97      0.97        30
           6       1.00      1.00      1.00        37
           7       0.97      0.97      0.97        37
           8       0.97      1.00      0.98        29
           9       0.91      0.94      0.93        34

    accuracy                           0.98       360
   macro avg       0.98      0.98      0.98       360
weighted avg       0.98      0.98      0.98       360



### 회고

------

머신러닝(, 딥러닝)은 노가다라고 교수님께 배웠다. 최적의 결과를 위해 하이퍼파라미터를 헤매는 과정은 답없는 싸움을 하는 것 같다. 
프로젝트에서 어려웠던 점은 역시 최적의 결과를 얻기위해 싸우는 과정이다. 해결을 위해 GridSearch방법을 사용해 시도해보았지만, 
많은 로드가 걸려 OS에서 프로세스를 종료시킨다. 그래서 되도록 많은 경우의 케이스를 고려하기위해 for문을 이용해 n_estimator와
min_samples_leaf 파라미터를 튜닝시키다 전부 accuracy가 98 % 여서 한계점이 있었다.

새로 배운 것은 프로젝트를 진행하면서 breast_cancer등의 샘플 데이터를 불러오는 것을 배웠다.



사실 하이퍼파라미터 튜닝은 우주전파재난환경 대회에서 수 시간동안 했던 일이었다. 
대회 측의 Weighted RMSE가 내가 직접계산하는 RMSE와 달랐던 탓에 최적의 결과를 구했는지는 아직도 의문이지만,
for문을 이용해서 최적의 하이퍼파라미터를 찾으려고 했던 치열했던 노력이 이 프로젝트를 진행하며 생각났다. 







#### 출처

------

- 랜덤포레스트 튜닝 : https://towardsdatascience.com/hyperparameter-tuning-the-random-forest-in-python-using-scikit-learn-28d2aa77dd74
- lazy predict : https://lazypredict.readthedocs.io/en/latest/